apiserver/httpcontext.go

```diff
@@ -104,9 +104,51 @@ func checkPermissions(tag names.Tag, acceptFunc common.GetAuthFunc) (bool, error
 	return false, errors.NotValidf("tag kind %v", tag.Kind())
 }
 
-// stateForRequestAuthenticatedUser is like stateForRequestAuthenticated
+// stateForMigration asserts that the incoming connection is from a user that
+// has admin permissions on the controller model. The method also gets the
+// model uuid for the model being migrated from a request header, and returns
+// the state instance for that model.
+func (ctxt *httpContext) stateForMigration(r *http.Request) (*state.State, error) {
+	st, user, err := ctxt.stateAndEntityForRequestAuthenticatedUser(r)
+	if err != nil {
+		return nil, err
+	}
+	if !st.IsController() {
+		return nil, errors.BadRequestf("model is not controller model")
+	}
+	admin, err := st.IsControllerAdmin(user.Tag().(names.UserTag))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if !admin {
+		return nil, errors.Unauthorizedf("not a controller admin")
+	}
+
+	modelUUID, err := validateModelUUID(validateArgs{
+		statePool: ctxt.srv.statePool,
+		modelUUID: r.Header.Get("Migration-Model-UUID"),
+		strict:    true,
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	migrationSt, err := ctxt.srv.statePool.Get(modelUUID)
```

Comment:
Thinking about it, I'm going to need to fix the other `stateForRequest*` functions in the same way as this, aren't I?